### DISCUSSION TITLE
Customization beyond Configuration.h

### DISCUSSION THREAD

**Discussion by billbrach** [aegean-odyssey/mpmd_marlin_1.1.x#66] [Q&A] [answered]
I have an older MPMD, and had to change the baud rate to 115200.  This was easy; I found the setting in Configuration.h.

Now, I want to change to the 5A power supply as I think I read somewhere that 10A was the default.  

I am not very familiar with the Linux/makefile system of building, as almost everything I've done like this has been done in the Arduino IDE on a Windows box.

I see this in Configuration.h but do not see where you set the var "ONLY_ONE_HEATER_AT_A_TIME"  I guess I could flip the 5Alimit and the 10Alimit but it seems there is some other way I don't know about.  Command line possibly ??

#if ONLY_ONE_HEATER_AT_A_TIME
#define __H_  "5Alimit"
#else
#define __H_  "10Alimit"
#endif

Thanks !!

Bill

**Comment by aegean-odyssey** [comment, marked as answer]
I'm a big proponent of explicitly documenting the build process for projects, especially embedded software. A "Makefile" is a relatively simple way to capture all of the details about how to compile, assemble, link and otherwise process the source code files. 

I've not tried to build the firmware with the Arduino tools, but under the hood, Arduino uses the same cross compiler tool chain, so I imagine that it is possible. 

First though, the baudrate setting has no bearing on the firmware. The USB port always operates at its fastest possible speed -- the baudrate is passed between the printer and the host more or less as a "faux" setting to appear like a "COMM" style serial port. I think the host computer ignores the speed setting as well when creating a virtual comm port via USB. 

Here are a few things to keep in mind:

* The makefile configures the build by defining a few preprocessor constants using the `-D` option of the `gcc` compiler. Providing one constant from each of the three groups below configures the build. It will be far simpler to add these definitions globally to the build (i.e. add three -D... to the compiler options) than to edit the source files everywhere. I'm pretty sure the Arduino tool will allow you to easily add these definitions.

```
SM0000 = -DMAKE_STEPPERS_0000
SM0001 = -DMAKE_STEPPERS_0001
SM1110 = -DMAKE_STEPPERS_1110
SM1111 = -DMAKE_STEPPERS_1111

AC_FAN = -DMAKE_AC_FAN
PC_FAN = -DMAKE_PC_FAN

L05AMP = -DMAKE_05ALIMIT
L10AMP = -DMAKE_10ALIMIT
```

* In a similar vein, you made need to adjust the compiler and linker options to build successfully. These options are documented in `Makefile` in the section labeled `### BUILD FLAGS`.

* _(Rather important)_ The project uses a build trick to minimize the changes to the original Marlin source files and to the original STM32 SDK source files. When compiling the Marlin source files, the file, `HAL_stm32.h`, is _ALWAYS_ included first using the `-include` option.  Likewise when compiling the SDK's (hal) source, the file,  `stm32f0xx_hal_conf.h`, is _ALWAYS_ included. Again, adding these options to Arduino's build process is likely necessary.

* _(Rather important)_ The project sets the paths to locate various included (header) files. These may differ from the paths that the Arduino tool normally configures. In particular the directory, `marlin_changes/arduino_fakes/` replaces  the true arduino header files to effectively remove AVR specific code and avoid changing the original Marlin source code.

The file, `Makefile`, should contain all of the information you need to adjust the Arduino tools. Also, with the Arduino tools installed (since it uses the same tool chain), it may be possible to just run 
```
make select='SM0000|SM0001|SM1110|SM1111  AC_FAN|PC_FAN  L05AMP|L10AMP'
``` 
You may need to edit the makefile to adjust the path to where the build tools are located -- this is in the `### BUILD TOOLS` section of the makefile.

> Note: the `mpmd_marlin_1.1.x` project requires only the [arm toolchain](
https://developer.arm.com/tools-and-software/open-source-software/developer-tools/gnu-toolchain/gnu-rm/downloads) and _"make"_ to build successfully -- in case you're looking for a minimal (local) build system.

Finally, if you're looking for a quick and easy way to build the firmware, you can build it directly on github by forking the project and using the provided github "action" to build a one-off version of the firmware. It is probably the easiest method.

**Reply by billbrach** [reply]
Sorry, I was not clear, I am NOT using the Arduino IDE, I am using your toolchain as suggested on the Development page of your repository.  I asked the question because I've never used Linux's Make command, except from within a supplied script while installing some software and everything is does scrolls by real fast LOL

Is your baud rate comment correct for OctoPrint, because that is what is connecting to my MPMD ??  I can watch the terminal commands and it looks like it is trying different baud rates when it cannot connect.  Before I flashed this firmware, it connected very quickly.  I had already re-flashed it with this firmware, set at the lower baud rate in Configuration.h, and it connected instantly as before.  It just occurred to me prior to trying to print, to check the power supply and that's why I posted the question.

So, if I understand your response, the options are added to the make command.

Something like this: 

"make -dOption1 -dOption2 -dOption3"
-or- 
"make SM0000 AC_FAN L05A" ??

BTW, your Development environment works great on Linux Mint 19.  Thanks for the quick response !!

**Comment by aegean-odyssey** [comment]
Hmm, the USB CDC protocol really doesn't care about baudrates. I suspect that both sides need to agree on a baudrate (eventhough the baudrate doesn't really matter) in order to establish the connection. It's possible that the negotiation for the 250000bps baudrate is somehow thwarted in your setup. No worries, the USB connection will run as fast as it possibly can regardless of the baudrate setting.

Glad to hear that you're using linux straight-away. I think that the Arduino IDE is overkill for this application.

The simplest way to use the makefile:

```sh
# make select='SM0000|SM0001|SM1110|SM1111  AC_FAN|PC_FAN  L05AMP|L10AMP'
# will generate the appropriate -D options for the gcc compiler.
# e.g.
make select='SM0001 AC_FAN L10AMP'
```

```sh
# hack-ish command to pull the build string out of the firmware image to confirm
grep -a -o 'mpmd[^)]*20..)' mpmd*.bin
```

Alternatively, you can modify the test build targets, `T1` and `T2`, in the file, `Makefile`, and build with `make all`. Btw, the project's travis-ci uses `make all` to automatically "test" build on every git push. It uses `make ALL` to create all sixteen (16) variants when building firmware for a release.
  
Excerpt from `Makefile`:
```
### testing variants

T1 : DEFINES += -DMAKE_05ALIMIT -DMAKE_AC_FAN -DMAKE_STEPPERS_0001
T1 = ${PROJECT}-${VERSION}r${RELEASE}-ACfan

T2 : DEFINES += -DMAKE_05ALIMIT -DMAKE_PC_FAN -DMAKE_STEPPERS_0001
T2 = ${PROJECT}-${VERSION}r${RELEASE}-PCfan
```

**Reply by billbrach** [reply]
Thanks, that worked.  Might you consider adding the make usage to the Development wiki ?? Might help the next guy LOL

**Comment by aegean-odyssey** [comment]
Agree on the wiki. It's probably time to overhaul the wiki and create a useful FAQs.
